kafka/integration_mock_test.go

```diff
@@ -0,0 +1,220 @@
+	assert.NoError(err, "Producer creation should succeed")
+	defer p.Close()
+	assert.NoError(err, "Message should be produced")
+
+	topic := "topic"
+	msg := Message{
+		TopicPartition: TopicPartition{
+			Topic:     &topic,
+			Partition: PartitionAny,
+		},
+		Value: []byte("value"),
+	}
+	err = p.Produce(&msg, nil)
+
+	cfg = &ConfigMap{
+		"bootstrap.servers": mockCluster.BootstrapServers(),
+		"group.id":          "group",
+	}
+
+	consumer, err := NewConsumer(cfg)
+	assert.NoError(err, "Consumer creation should succeed")
+
+	partitionsRevoked := false
+	err = consumer.SubscribeTopics([]string{topic}, func(c *Consumer, e Event) error {
+		switch e.(type) {
+		case RevokedPartitions:
+			partitionsRevoked = true
+			_, err := c.Commit()
+			assert.NotErrorIs(err, getOperationNotAllowedErrorForClosedClient(),
+				"Commit should not fail with '%s'",
+				getOperationNotAllowedErrorForClosedClient())
+		}
+		return nil
+	})
+
+	// Poll for enough time that we do the rebalance.
+	consumer.Poll(10 * 1000)
+
+	err = consumer.Close()
+	assert.NoError(err, "Consumer closure should succeed")
+	assert.True(partitionsRevoked, "Partitions should be revoked after close.")
+}
+
+// TestParallelConsumerCloses tests that two goroutines cannot close the
+// consumer at one time.
+func TestParallelConsumerCloses(t *testing.T) {
+	assert := assert.New(t)
+
+	// Create mock cluster
+	mockCluster, err := NewMockCluster(2)
+	assert.NoError(err, "Mock cluster creation should succeed")
+	defer mockCluster.Close()
+
+	// Create topic by producing to it.
+	// TODO: Later, use CreateTopic, once that PR is merged.
+	cfg := &ConfigMap{
+		"bootstrap.servers": mockCluster.BootstrapServers(),
+	}
+	p, err := NewProducer(cfg)
+	assert.NoError(err, "Producer creation should succeed")
+	defer p.Close()
+	assert.NoError(err, "Message should be produced")
+
+	topic := "topic"
+	msg := Message{
+		TopicPartition: TopicPartition{
+			Topic:     &topic,
+			Partition: PartitionAny,
+		},
+		Value: []byte("value"),
+	}
+	err = p.Produce(&msg, nil)
+
+	cfg = &ConfigMap{
+		"bootstrap.servers": mockCluster.BootstrapServers(),
+		"group.id":          "group",
+	}
+
+	consumer, err := NewConsumer(cfg)
+	assert.NoError(err, "Consumer creation should succeed")
+
+	err = consumer.SubscribeTopics([]string{topic}, func(c *Consumer, e Event) error {
+		switch e.(type) {
+		case RevokedPartitions:
+			// Slow down the closure, since the close method will wait for this
+			// callback to run first.
+			time.Sleep(time.Second)
+		}
+		return nil
+	})
+
+	// Poll for enough time that we do the rebalance.
+	consumer.Poll(10 * 1000)
+
+	// Try closing the consumer from two different go-routines
+	var err0 error
+	var err1 error
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go func() {
+		err0 = consumer.Close()
+		wg.Done()
+	}()
+	go func() {
+		err1 = consumer.Close()
+		wg.Done()
+	}()
+
+	wg.Wait()
+
+	assert.True((err0 == nil) || (err1 == nil), "Exactly one of the errors must be nil")
```

Comment:
Can use a single condition
```suggestion
	assert.True((err0 == nil) != (err1 == nil), "Exactly one of the errors must be nil")
```